rust-runtime/aws-smithy-runtime/src/client/waiters.rs

```diff
@@ -0,0 +1,517 @@
+use aws_smithy_runtime_api::client::{
+    result::CreateUnhandledError,
+    waiters::error::{ExceededMaxWait, FailureState, OperationFailed, WaiterError},
+};
+use std::future::Future;
+use std::time::Duration;
+
+mod backoff;
+
+/// Waiter acceptor state
+///
+/// This enum (vaguely) matches the [acceptor state] from the Smithy spec.
+/// It has an additional `NoAcceptorsMatched` variant to indicate the case where
+/// none of the modeled waiters matched the response, which the spec mentions but
+/// doesn't consider an official part of the acceptor state enum. An `Option<AcceptorState>`
+/// could have been used instead, but this seemed cleaner.
+///
+/// [acceptor state]: https://smithy.io/2.0/additional-specs/waiters.html#acceptorstate-enum
+#[non_exhaustive]
+#[derive(Copy, Clone, Debug, Eq, PartialEq)]
+pub enum AcceptorState {
+    /// None of the modeled acceptors matched the response.
+    NoAcceptorsMatched,
+    /// A `success` acceptor matched the response.
+    Success,
+    /// A `failure` acceptor matched the response.
+    Failure,
+    /// A `retry` acceptor matched the response.
+    Retry,
+}
+
+/// Orchestrates waiting via polling with jittered exponential backoff.
+///
+/// This is meant to be used internally by the generated code to provide
+/// waiter functionality.
+pub struct WaiterOrchestrator<AcceptorFn, OperationFn> {
+    backoff: Backoff,
+    time_source: SharedTimeSource,
+    sleep_impl: SharedAsyncSleep,
+    acceptor_fn: AcceptorFn,
+    operation_fn: OperationFn,
+}
+
+impl WaiterOrchestrator<(), ()> {
+    /// Returns a builder for the waiter orchestrator.
+    pub fn builder() -> WaiterOrchestratorBuilder<(), ()> {
+        WaiterOrchestratorBuilder::default()
+    }
+}
+
+impl<AcceptorFn, OperationFn> WaiterOrchestrator<AcceptorFn, OperationFn> {
+    fn new(
+        backoff: Backoff,
+        time_source: SharedTimeSource,
+        sleep_impl: SharedAsyncSleep,
+        acceptor_fn: AcceptorFn,
+        operation_fn: OperationFn,
+    ) -> Self {
+        WaiterOrchestrator {
+            backoff,
+            time_source,
+            sleep_impl,
+            acceptor_fn,
+            operation_fn,
+        }
+    }
+}
+
+impl<AcceptorFn, OperationFn, O, E, Fut> WaiterOrchestrator<AcceptorFn, OperationFn>
+where
+    AcceptorFn: Fn(Result<&O, &E>) -> AcceptorState,
+    OperationFn: Fn() -> Fut,
+    Fut: Future<Output = Result<O, SdkError<E, HttpResponse>>>,
+    E: CreateUnhandledError + std::error::Error + Send + Sync + 'static,
+{
+    /// Orchestrates waiting via polling with jittered exponential backoff.
+    pub async fn orchestrate(
+        self,
+    ) -> Result<FinalPoll<O, SdkError<E, HttpResponse>>, WaiterError<O, E>> {
+        let start_time = self.time_source.now();
+        let mut attempt = 0;
+        let mut done_retrying = false;
+        loop {
+            tracing::debug!("executing waiter poll attempt #{}", attempt + 1);
```

Comment:
**question**: What do the logs look like for a single waiter execution with multiple attempts? Each attempt will be a separate operation which (after we add a unique ID for each invoke) will result in each attempt being uncorrelated to the original waiter logs. Can we do better?

Reply:
I had a hunch you'd comment about this. I'll take a look at what we can do to correlate them.

Reply:
I attached a correlation ID in e1e103fb672dc916f8f380fd50ec306b90a722ff.

Looks like the following in the logs:

```
2024-04-24T23:12:44.983787Z DEBUG waiter{waiter_id=5652710}: aws_smithy_runtime::client::waiters: executing waiter poll attempt #1
2024-04-24T23:12:44.984031Z DEBUG waiter{waiter_id=5652710}:invoke{service=ec2 operation=DescribeInstanceStatus}:apply_configuration: aws_smithy_runtime::client::orchestrator: timeout settings for this operation: TimeoutConfig { connect_timeout: Set(3.1s), read_timeout: Disabled, operation_timeout: Disabled, operation_attempt_timeout: Disabled }
...
2024-04-24T23:13:00.268964Z DEBUG waiter{waiter_id=5652710}: aws_smithy_runtime::client::waiters: delay calculated for attempt #2; elapsed (15.285171s); waiter will poll again in 24s
```